language/stdlib/modules/Roles.move

```diff
@@ -285,6 +288,76 @@ module Roles {
         move_to(new_account, Privilege<UnhostedRole>{ witness: UnhostedRole{}, is_extracted: false })
     }
 
+    // DD refactor: privilege-checking functions for roles
+    /// assert_has_*_role and assert_has_*_privilege functions are derived exactly from the
+    /// "Libra Governance roles & permissions" spreadsheet
+    /// https://docs.google.com/spreadsheets/d/13pL_0JCdNkGTMemVYKLvGifiF-svP_fW_pT6CjxFMCA/edit#gid=0
+    /// MintCurrency, BurnCurrency, PreburnCurrency, RotateAuthenticationKey, WithdrawalCapability are
+    /// omitted because they all involve real capabilities. They were also not handled in the
+    /// Roles module before refactoring.
+    ///
+    /// Naming conventions: Many of the "has_*_privilege" functions do have the same body
+    /// because the spreadsheet grants all such privileges to addresses (usually a single
+    /// address) with that role. In effect, having the privilege is equivalent to having the
+    /// role, but the function names document the specific privilege involved.  Also, modules
+    /// that use these functions as a privilege check can hide specific roles, so that a change
+    /// in the privilege/role relationship can be implemented by changing Roles and not the
+    /// module that uses it.
+
+    public fun assert_has_role(account: &signer, role_id: u64) acquires RoleId {
```

Comment:
As mentioned/explained earlier, please do not add `assert_` functions like this--the `assert` needs to be at the call site so we can see what was happening when the privilege check fails.

Reply:
I was the one suggesting to introduce this, and I did already something similar in submitted code (`LibraTimestamp::assert_is_genesis`). This is ugly. The reason why we are doing this is to ensure error codes are consistently used. The error code story is kind of a big mess right now. I don't quite understand why Move VM wasn't designed from the beginning to give you a stack trace on abort, it is a very basic feature.

If we roll this back, we should also do other places, and think about at least making error codes constants throughout the place (module `ErrorCodes`).

Reply:
Totally agreed on rolling it back in other places + making all error codes constants. This is planned (https://github.com/orgs/libra/projects/7), but things are indeed a mess right now (we were only recently unblocked by source lang constants). Just don't want to make the problem worse before we fix it :). 

>  I don't quite understand why Move VM wasn't designed from the beginning to give you a stack trace on abort, it is a very basic feature.

This is something that we have (though it's not enabled automatically), but the problem is that we can't use this on-chain. The `vm_status` part of a transaction output needs to be constant size and fairly small. Today, you only get `SUCCESS` or `ABORTED` (not even the abort error code!). @tnowacki and @tzakian are working on adding at least the module name and abort error code to mitigate this.

Reply:
If you have this but can't use on-chain I'd say this is a PITA for DevOps. If you have an unexpected failure in production, you want to be able to identify its source and mitigate as fast as possible. This should strictly override concerns about resource usage (and has in production environments I've seen before).

@dariorussi

Reply:
This is a nuanced topic that I probably shouldn't have brought up on in a GH comment thread, but let me explain a few more of the nuances since we're here:

- There are two ways to get a stack trace and other metadata from an aborted execution: (1) in the VM logs (2) the `TransactionInfo` stored directly on the blockchain. We have very good support for (1); the logs are thorough and there is tooling for accessing, scraping, and indexing these logs. This is the way DevOps folks running nodes should/will get at VM information (and totally agreed that stack traces and other data is essential there.
- (2) is how clients will get data about aborted execution (since they don't have access to the logs of the validator node). There are definite tradeoffs on having more vs less information here. Resource usage is one concern, but not the only or primary one. Another concern is maintaining a small TCB and spec footprint. Note that the `vm_status` part of an executed transaction is stored on the blockchain, so changes there affect the Merkle root hash. If the status is more complex (e.g., a stack trace), the code for producing that stack trace is in the TCB and the format of that stack trace is part of the Libra spec. 
- The reasoning behind including only a status code and module name name in (2) is that it strikes a reasonable balance between client friendliness and TCB/resource usage/other concerns. A client gets decent info by reading the blockchain, and it can (if necessary) get the full Libra logs by syncing to the latest blockchain state and replaying the transaction execution locally.

Reply:
I'm using individual asserts now, based on an earlier comment.  It's not too hard to change back if we want to.